### PR TITLE
Improve user status revert performance

### LINKED ISSUE #31190: Call to `setUserStatus` does too many database queries
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Calling `OCP\UserStatus\IManager::setUserStatus` does way too many database queries

### Steps to reproduce

1. Start a call in Talk
2. See that your status is set to "In a call"

### Expected behavior

Only very few queries executed.

### Installation method

_No response_

### Operating system

_No response_

### PHP engine version

_No response_

### Web server

_No response_

### Database engine version

_No response_

### Is this bug present after an update or on a fresh install?

_No response_

### Are you using the Nextcloud Server Encryption module?

_No response_

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

```shell
-
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by nickvergessen**
Currently if a meeting in Talk is ended for everyone the database basically dies. One of the problems is resetting the user status:

### Queries before "5 per user"
1. Get user backup status `revertUserStatus`
2. Get user status `revertUserStatus`
3. Get user status `removeUserStatus`
4. Delete user status `removeUserStatus`
5. Update user backup status `revertUserStatus`

### Plan level 1: Improve current function: "3 per user"
- [x] Get user backup status `revertUserStatus`: ☑️
- [x] Get user status `revertUserStatus`: 💥 Remove by adding a delete which has the matching parameters
- [x] Get user status `removeUserStatus`: 💥 Removed in this PR
- [x] Delete user status `removeUserStatus`: ☑️
- [x] Update user backup status `revertUserStatus`: ☑️

### Plan level 2: Add special handling for "End call for everyone": "3 in total"
- [x] Get the user + backup status of all users in single query
- [x] Delete statuses by id in single query
- [x] Update statuses in single query

### Plan level 3
- [x] There is currently no index on is_backup so all those queries might do a `using WHERE` => removed from query as the leading underscore ensures it's value already

Fix #31190

### CODE REVIEW

**Comment by nickvergessen** [comment]
at least the first/first-two commits should be backported to 23

File: apps/user_status/lib/Service/StatusService.php [hunk 1]
```diff
@@ -460,27 +461,52 @@ public function backupCurrentStatus(string $userId): bool {
...
+				$backups[$userStatus->getUserId()] = $userStatus->getId();
+			}
+		}
+
+		// For users with both delete the status when matching
+		foreach ($statuesToDelete as $userId => $statusId) {
+			if (isset($backups[(string) $userId])) {
+				$restoreIds[] = $backups[(string) $userId];
+			}
+		}
```
Comment (nickvergessen) [inline apps/user_status/lib/Service/StatusService.php:505]: Need to investigate, in theory the single user method handles a case where the user has a backup but no actual current status.

**Review by CarlSchwan — APPROVED** [review]

**Comment by nickvergessen** [comment]
3 way comparison in talks integration test of the query log makes it obvious:
Sample: `./run.sh features/callapi/group.feature`

left before | middle with this PR | right with user status disabled
---|---|---
2884 | 2594 | 2505

Bildschirmfoto von 2022-02-16 15-46-16

**Review by CarlSchwan — APPROVED** [review]

**Comment by vitormattos** [comment]
Drone tests isn't necessary?

File: apps/user_status/lib/Service/StatusService.php [hunk 2]
```diff
@@ -434,53 +491,71 @@ private function addDefaultMessage(UserStatus $status): void {
...
+			$findById[] = '_' . $userId;
+		}
+		$userStatuses = $this->mapper->findByUserIds($findById);
+
+		$backups = $restoreIds = $statuesToDelete = [];
+		foreach ($userStatuses as $userStatus) {
+			if (!$userStatus->getIsBackup()
+				&& $userStatus->getMessageId() === $messageId
+				&& $userStatus->getStatus() === $status) {
+				$statuesToDelete[$userStatus->getUserId()] = $userStatus->getId();
```
Comment (PVince81) [inline apps/user_status/lib/Service/StatusService.php:542]: typo deleting statues :statue_of_liberty:

**Review by PVince81 — APPROVED** [review]
:+1:

**Pull request merged**